File: practice/src/main/kotlin/ar/com/miura/jvm/string/GroupAnagrams.java
package ar.com.miura.jvm.string;

public class GroupAnagrams {
    public java.util.List<java.util.List<String>> groupAnagrams(String[] strs) {

        java.util.Map<String, java.util.List<String>> map = new java.util.HashMap<>();

        for(String string:strs) {
            char[] chars = string.toCharArray();
            java.util.Arrays.sort(chars);
            String key = new String(chars);

            java.util.List<String> words = map.getOrDefault(key, new java.util.ArrayList<String>());
            words.add(string);
            map.put(key, words);
        }

        java.util.List<java.util.List<String>> words = new java.util.ArrayList<>();

        for(java.util.Map.Entry<String, java.util.List<String>> entry:map.entrySet()) {
            java.util.List<String> list = entry.getValue();
            words.add(list);
        }

        return words;

    }
}

File: practice/src/main/kotlin/ar/com/miura/lc/array/GroupAnagrams.kt
package ar.com.miura.lc.array

/**
 * https://leetcode.com/explore/featured/card/top-interview-questions-medium/103/array-and-strings/778/
 */
class GroupAnagrams {
    fun groupAnagrams(strs: Array<String>): List<List<String>> {
        val map = mutableMapOf<String, MutableList<String>>()

        for ((k, word) in strs.withIndex()) {
            var charArray = word.toCharArray()
            charArray.sort()
            var key = String(charArray)

            var words = map.getOrDefault(key, mutableListOf<String>())
            words.add(word)
            map.put(key, words)
        }

        var anagrams = mutableListOf<List<String>>()
        for ((k, v) in map) {
            anagrams.add(v)
        }

        return anagrams.sortedWith(compareBy { it.size })
    }
}